zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -1148,7 +1148,8 @@ private void startConnect(InetSocketAddress addr) throws IOException {
                     if (zooKeeperSaslClient != null) {
                         zooKeeperSaslClient.shutdown();
                     }
-                    zooKeeperSaslClient = new ZooKeeperSaslClient(SaslServerPrincipal.getServerPrincipal(addr, clientConfig), clientConfig);
+                    zooKeeperSaslClient = new ZooKeeperSaslClient(clientConfig);
+                    zooKeeperSaslClient.createSaslClient(SaslServerPrincipal.getServerPrincipal(addr, clientConfig));
```

Comment:
What's the point of this change?
You split the logic into 2 methods which you call one after the other. I don't see how this makes the code more resource efficient.

Reply:
The benefit is that you will never see the constructor fail and then you can have a handle to perform cleanup

Reply:
Got it.